src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -0,0 +1,183 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This portion of the binder converts an <see cref="ExpressionSyntax"/> into a <see cref="BoundExpression"/>.
+    /// </summary>
+    internal partial class Binder
+    {
+        private BoundExpression BindWithExpression(WithExpressionSyntax syntax, DiagnosticBag diagnostics)
+        {
+            // PROTOTYPE: this entire method is likely to change
+            var receiver = BindRValueWithoutTargetType(syntax.Receiver, diagnostics);
+            var receiverType = receiver.Type;
+
+            var lookupResult = LookupResult.GetInstance();
+            HashSet<DiagnosticInfo>? useSiteDiagnostics = null;
+            bool hasErrors = false;
+
+            var args = ArrayBuilder<(Symbol?, BoundExpression)>.GetInstance();
+            // Bind with expression arguments
+            foreach (var initializer in syntax.Initializers)
+            {
+                var propName = initializer.NameEquals?.Name.Identifier.Text;
+                Symbol? member = null;
+                if (!(propName is null) && !(receiverType is null))
+                {
+                    var location = initializer.NameEquals!.Name.Location;
+                    this.LookupMembersInType(
+                        lookupResult,
+                        receiverType,
+                        propName,
+                        arity: 0,
+                        basesBeingResolved: null,
+                        options: LookupOptions.Default,
+                        originalBinder: this,
+                        diagnose: false,
+                        useSiteDiagnostics: ref useSiteDiagnostics);
+                    if (lookupResult.IsSingleViable &&
+                        lookupResult.SingleSymbolOrDefault is var sym)
+                    {
+                        switch (sym.Kind)
+                        {
+                            case SymbolKind.Property:
+                                // PROTOTYPE: this should check for init-only, but that isn't a separate feature yet
+                                if (!(sym is SynthesizedRecordPropertySymbol))
```

Comment:
This currently means we only support records in the same compilation, correct? Just making sure I understand the current limitations

Reply:
That's correct.